src/transformers/data/data_collator.py

```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+                raise ValueError(
+                    "You are attempting to pad samples but the tokenizer you are using"
+                    f" ({self.tokenizer.__class__.__name__}) does not have one."
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The mask is determined by iteratively performing the following steps until the end of the sequence:
+            1. Sample an n-gram length (n) from [1, ..., max_gram] (L in the paper)
+            2. Reserve a context of length of n-gram length / mlm_probability (In the paper, context size is K*n, where K is close to the inverse of mlm_probability)
+            3. Sample a starting point and mask n tokens continuously from then
+        """
+
+        if self.tokenizer.mask_token is None:
+            raise ValueError(
+                "This tokenizer does not have a mask token which is necessary for masked language modeling. Remove the --mlm flag if you want to use this tokenizer."
+            )
+
+        labels = inputs.clone()
+        # Creating the mask and target mapping tensors
+        masked_indices = torch.full(labels.shape, 0, dtype=torch.bool)
+        target_mapping = torch.zeros((labels.size(0), labels.size(1), labels.size(1)), dtype=torch.float)
+
+        for i in range(labels.size(0)):
+            cur_len = 0
+            max_len = labels.size(1)
+            # Since we're replacing non-masked (non-functional) tokens with -100 in the labels tensor,
+            # the i-th label corresponds to the i-th token.
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
+                start = cur_len + torch.randint(ctx_size - n + 1, (1,)).item()
+                end = start + 1
+                while end < max_len and end < start + n:
+                    end += 1
+                masked_indices[i, start:end] = 1
+                cur_len += ctx_size
+
+        special_tokens_mask = torch.tensor(
+            [self.tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()],
+            dtype=torch.bool,
+        )
+        masked_indices.masked_fill_(special_tokens_mask, value=0.0)
+        if self.tokenizer._pad_token is not None:
+            padding_mask = labels.eq(self.tokenizer.pad_token_id)
+            masked_indices.masked_fill_(padding_mask, value=0.0)
+
+        # Mask indicating non-functional tokens, where functional tokens are [SEP], [CLS], padding, etc.
+        non_func_mask = ~(padding_mask & special_tokens_mask)
+
+        inputs[masked_indices] = self.tokenizer.convert_tokens_to_ids(self.tokenizer.mask_token)
```

Comment:
better to use `self.tokenizer.mask_token_id` instead of calling `convert_tokens_to_ids`